amqp/src/test/java/io/cloudevents/amqp/ProtonBasedAmqpMessageWriterTest.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.cloudevents.amqp;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.qpid.proton.amqp.Binary;
+import org.apache.qpid.proton.amqp.messaging.ApplicationProperties;
+import org.apache.qpid.proton.amqp.messaging.Section;
+import org.apache.qpid.proton.message.Message;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.amqp.impl.ProtonBasedAmqpMessageWriter;
+import io.cloudevents.core.format.EventFormat;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.mock.CSVFormat;
+import io.cloudevents.core.test.Data;
+import io.cloudevents.types.Time;
+
+/**
+ * Tests verifying the behavior of the {@code ProtonBasedAmqpMessageWriter}.
+ */
+public class ProtonBasedAmqpMessageWriterTest {
+
+    /**
+     * Verifies that a binary CloudEvent message can be successfully represented
+     * as an AMQP message.
+     */
+    @Test
+    public void testWriteBinaryCloudEventToAmqpRepresentation() {
+
+        final CloudEvent binaryEvent = Data.V1_WITH_JSON_DATA;
+        final Message expectedMessage = translateBinaryEvent(binaryEvent);
+
+        final MessageWriter<?, Message> writer = new ProtonBasedAmqpMessageWriter<Message>();
+        final Message actualMessage = writer.writeBinary(binaryEvent);
+
+        assertThat(actualMessage.getContentType()).isEqualTo(expectedMessage.getContentType());
+        assertThat(actualMessage.getBody().toString()).isEqualTo(expectedMessage.getBody().toString());
+        assertThat(actualMessage.getApplicationProperties().getValue()).
+                  isEqualTo(expectedMessage.getApplicationProperties().getValue());
+    }
+
+    /**
+     * Verifies that a structured CloudEvent message (in CSV format) can be successfully represented
+     * as an AMQP message.
+     */
+    @ParameterizedTest()
+    @MethodSource("io.cloudevents.core.test.Data#allEventsWithoutExtensions")
+    public void testWriteStructuredCloudEventToAmqpRepresentation(final CloudEvent event) {
+        final EventFormat format = CSVFormat.INSTANCE;
+        final Message expectedMessage = translateStructured(event, format);
+
+        final MessageWriter<?, Message> writer = new ProtonBasedAmqpMessageWriter<Message>();
+        final Message actualMessage = writer.writeStructured(event, format.serializedContentType());
+
+        assertThat(actualMessage.getContentType()).isEqualTo(expectedMessage.getContentType());
+        assertThat(actualMessage.getBody().toString()).isEqualTo(expectedMessage.getBody().toString());
+        assertThat(actualMessage.getApplicationProperties()).isNull();
+    }
+
+    private Message translateBinaryEvent(final CloudEvent event) {
```

Comment:
Can you try a message with cloudevents extensions too?

Reply:
i notice a small bug(?) while testing this. 

In `io.cloudevents.SpecVersion`, the `datacontentencoding` attribute name is included in the list of optional attributes for the V03 (i.e in CloudEventAttributes#getAttributeNames()). However, when i use that name to retrieve its value using CloudEventV03#getAttribute(String), i get an `IllegalArgumentException:  Spec version v0.3 doesn't have attribute named datacontentencoding`;

i think an additional branch should be added to the switch statement to handle the case when the attributeName is `datacontentencoding`. Since the value of this attribute is not stored inside a CloudEventV03 object, we can just do nothing in this case, instead of raising an exception. any thoughts, @slinkydeveloper ? this is what am doing now to make my tests happy 😄

Reply:
The `datacontentencoding` is a funny one :smile: In the 0.3 spec, it was a bit unclear its role outside the json itself. So in this sdk, I decided to don't save it in the event, because it doesn't make sense to do so. It is in fact only used in the json format, when the event is marshalled/unmarshalled back and forth the wire.

I guess the right behaviour here is that, when `CloudEventV03#getAttribute(String)` is invoked with `datacontentencoding`, it should just return null always.

Reply:
I can provide a patch for it, one moment

Reply:
Here it is https://github.com/cloudevents/sdk-java/pull/279 Mind doing a review?

Reply:
ok, done